andela/ah-infinity-stones
# #161966917 Implement article rating functionality

#### What does this PR do?
- Implement article rating functionality.
#### Description of Task to be completed?
- Write article rating tests. They should be failing.
- Create an endpoint for rating an individual article
- Implement article rating functionality
- Make the tests pass.
#### How should this be manually tested?
1. `git clone -b ft-users-able-rate-article-161966917 https://github.com/andela/ah-infinity-stones` and `$ cd ah-infinity-stones`
2. Create and activate virtual env
- `virtualenv  -p python3 venv`
- `$ source venv/bin/activate`
3. `$ pip3 install -r requirements.txt` to install the dependencies
4. Setup environment variables. Copy all from below and paste them into the .env file. Update the values to your choice.
```export DATABASE_NAME='authors_haven'
export DATABASE_USER='postgres'
export DATABASE_PASSWORD='postgres'
export HOST='localhost'
export PORT='5432'
export DATABASE_TEST='test_authors_haven'
export EMAIL_HOST_USER='Infinites'
export EMAIL_HOST_PASSWORD='st1fflers'
export SECRET_KEY='my secret my choice'
```
Activate the env variables by `$source .env`.
5. Run migrations: `$ python3 manage.py migrate`
6. Run app - `$ python3 manage.py runserver`
7. Acess `http://127.0.0.1:8000/api/users/` through [Postman](https://www.getpostman.com/), provide inputs in the following format and create account.
```
{ "user": {
"username": "your name",
"email": "your@email.com",
"password": "@Password321"	
}
}
```
Check your email and activate the account. Use the registration token to access the following endpoints:
- `http://127.0.0.1:8000/api/articles` - Post an article. 
```
{
    "title": "How to make it rain",
    "description": "Ever wonder how?",
    "body": "You have to pray",
    "tag": ["reactjs", "angularjs", "rain"]
  
}
```
-  To rate an article, copy and paste the slug ( in this case `how-to-make-it-rain`) into the url and make a POST request to `http://127.0.0.1:8000/api/articles/<slug>/rate` (eg. ` http://127.0.0.1:8000/api/articles/how-to-make-it-rain/rate` ) with the following data:
```
{
    "rating": "4"
}
```

- To view all the ratings for a particular post use the same url but make a GET request.
      `http://127.0.0.1:8000/api/articles/<slug>/rate` (eg. ` http://127.0.0.1:8000/api/articles/how-to-make-it-rain/rate` ) 
- In order to view the article's rating average simply visit the following url(not forgetting your slug).
`http://127.0.0.1:8000/api/articles/<slug>` (eg. ` http://127.0.0.1:8000/api/articles/how-to-make-it-rain` ) 

8. Run tests`python3 manage.py test authors.apps.articles.tests.test_views.ArticleRatingTestCase`
#### What are the relevant pivotal tracker stories?
[#161966917](https://www.pivotaltracker.com/story/show/161966917)
#### Screenshots
<img width="1125" alt="screenshot 2018-12-19 at 20 58 26" src="https://user-images.githubusercontent.com/16904946/50252083-f217e080-03f6-11e9-8a02-bc08829acf73.png">
<img width="1118" alt="screenshot 2018-12-19 at 21 00 46" src="https://user-images.githubusercontent.com/16904946/50252086-f3e1a400-03f6-11e9-8f02-e3088e2b8abb.png">
<img width="1119" alt="screenshot 2018-12-19 at 21 01 05" src="https://user-images.githubusercontent.com/16904946/50252092-f9d78500-03f6-11e9-8f6c-d834b10eee95.png">
<img width="1123" alt="screenshot 2018-12-19 at 21 01 36" src="https://user-images.githubusercontent.com/16904946/50252096-ff34cf80-03f6-11e9-8367-d6cd61d92459.png">
<img width="1120" alt="screenshot 2018-12-19 at 23 27 33" src="https://user-images.githubusercontent.com/16904946/50252099-01972980-03f7-11e9-9c35-459fc700599e.png">
<img width="997" alt="screenshot 2018-12-19 at 20 33 54" src="https://user-images.githubusercontent.com/16904946/50252105-052ab080-03f7-11e9-99c0-173f8860ef24.png">
